src/mito2/src/read/flat_projection.rs

```diff
@@ -0,0 +1,221 @@
+        // TODO(yingwen): Support different flat schema options.
+        let format_projection = FormatProjection::compute_format_projection(
+            &id_to_index,
+            // All columns with internal columns.
+            metadata.column_metadatas.len() + 3,
+            column_ids.iter().copied(),
+        );
+
+        let batch_indices: Vec<_> = column_ids
+            .iter()
+            .map(|id| {
+                // Safety: The map is computed from `projection` itself.
+                format_projection
+                    .column_id_to_projected_index
+                    .get(id)
+                    .copied()
+                    .unwrap()
+            })
+            .collect();
+
+        let batch_schema = plain_projected_columns(metadata, &format_projection);
+
+        Ok(FlatProjectionMapper {
+            metadata: metadata.clone(),
+            output_schema,
+            column_ids,
+            batch_schema,
+            is_empty_projection,
+            batch_indices,
+        })
+    }
+
+    /// Returns a new mapper without projection.
+    pub fn all(metadata: &RegionMetadataRef) -> Result<Self> {
+        FlatProjectionMapper::new(metadata, 0..metadata.column_metadatas.len())
+    }
+
+    /// Returns the metadata that created the mapper.
+    pub(crate) fn metadata(&self) -> &RegionMetadataRef {
+        &self.metadata
+    }
+
+    /// Returns ids of projected columns that we need to read
+    /// from memtables and SSTs.
+    pub(crate) fn column_ids(&self) -> &[ColumnId] {
+        &self.column_ids
+    }
+
+    /// Returns ids of columns of the batch that the mapper expects to convert.
+    #[allow(dead_code)]
+    pub(crate) fn batch_schema(&self) -> &[(ColumnId, ConcreteDataType)] {
+        &self.batch_schema
+    }
+
+    /// Returns the schema of converted [RecordBatch].
+    /// This is the schema that the stream will output. This schema may contain
+    /// less columns than [ProjectionMapper::column_ids()].
+    pub(crate) fn output_schema(&self) -> SchemaRef {
+        self.output_schema.clone()
+    }
+
+    /// Returns an empty [RecordBatch].
+    pub(crate) fn empty_record_batch(&self) -> RecordBatch {
+        RecordBatch::new_empty(self.output_schema.clone())
+    }
+
+    /// Converts a flat format [RecordBatch] to a normal [RecordBatch].
+    ///
+    /// The batch must match the `projection` using to build the mapper.
+    #[allow(dead_code)]
+    pub(crate) fn convert(
+        &self,
+        batch: &datatypes::arrow::record_batch::RecordBatch,
+    ) -> common_recordbatch::error::Result<common_recordbatch::RecordBatch> {
+        if self.is_empty_projection {
+            return RecordBatch::new_with_count(self.output_schema.clone(), batch.num_rows());
+        }
+
+        let mut columns = Vec::with_capacity(self.output_schema.num_columns());
+        for index in &self.batch_indices {
+            let array = batch.column(*index).clone();
```

Comment:
Is there any case that input `batch` does not have array at `index`?

Reply:
Then this is an unexpected case and should be a bug.